exec/java-exec/src/main/java/org/apache/drill/exec/store/log/LogRecordReader.java

```diff
@@ -0,0 +1,261 @@
+    timeFormat = config.getTimeFormat();
+    errorOnMismatch = config.getErrorOnMismatch();
+
+    /*
+    This section will check for;
+    1.  Empty regex
+    2.  Invalid Regex
+    3.  Empty date string if the date format is used
+    4.  No capturing groups in the regex
+    5.  Incorrect number of data types
+    6.  Invalid data types
+     */
+    if (regex.isEmpty()) {
+      throw UserException.parseError().message("Log parser requires a valid, non-empty regex in the plugin configuration").build(logger);
+    } else {
+      //TODO Check for invalid regex
+      r = Pattern.compile(regex);
+      Matcher m = r.matcher("test");
+      if (m.groupCount() == 0) {
+        throw UserException.parseError().message("Invalid Regular Expression: No Capturing Groups", 0).build(logger);
+      } else if (m.groupCount() != (fieldNames.size())) {
+        throw UserException.parseError().message("Invalid Regular Expression: Field names do not match capturing groups.  There are " + m.groupCount() + " captured groups in the data and " + fieldNames.size() + " specified in the configuration.", 0).build(logger);
+
+      } else if ((dataTypes == null) || m.groupCount() != dataTypes.size()) {
+        //If the number of data types is not correct, create a list of varchar
+        dataTypes = new ArrayList<String>();
+        for (int i = -0; i < m.groupCount(); i++) {
+          dataTypes.add("VARCHAR");
+        }
+      }
+    }
+
+    //Check and set up date formats
+    if (dataTypes.contains("DATE") || dataTypes.contains("TIMESTAMP")) {
+      if (dateFormat != null && !dateFormat.isEmpty()) {
+        df = new java.text.SimpleDateFormat(dateFormat);
+      } else {
+        throw UserException.parseError().message("Invalid date format.  The date formatting string was empty.  Please specify a valid date format string in the configuration for this data source.", 0).build(logger);
+      }
+    }
+
+    if (dataTypes.contains("TIME")) {
+      if (timeFormat != null && !timeFormat.isEmpty()) {
+        tf = new java.text.SimpleDateFormat(timeFormat);
+      } else {
+        throw UserException.parseError().message("Invalid time format.  The time formatting string was empty.  Please specify a valid time format string in the configuration for this data source.", 0).build(logger);
+      }
+    }
+
+  }
+
+  public int next() {
+    this.writer.allocate();
+    this.writer.reset();
+
+    int recordCount = 0;
+
+    try {
+      BaseWriter.MapWriter map = this.writer.rootAsMap();
+      String line = null;
+
+      while (recordCount < MAX_RECORDS_PER_BATCH && (line = this.reader.readLine()) != null) {
+        lineCount++;
+
+        // Skip empty lines
+        if (line.trim().length() == 0) {
+          continue;
+        }
+
+        this.writer.setPosition(recordCount);
+        map.start();
+
+        Matcher m = r.matcher(line);
+        if (m.find()) {
+          for (int i = 1; i <= m.groupCount(); i++) {
```

Comment:
Again, the computer will handle this just fine if we move the row-processing code into another method. But, it will reduce the nesting levels us poor reviewers have to track.